### PR TITLE
Bug 1740035: Azure should create custom container

### PR DESCRIPTION
**Pull request by coreydaley**
Azure should create a custom container with name provided by user
Also fixed up a few other parts of the Azure CreateStorage flow

### CODE REVIEW

File: pkg/storage/azure/azure.go [hunk 1]
```diff
@@ -360,13 +354,26 @@ func (d *driver) StorageExists(cr *imageregistryv1.Config) (bool, error) {
 	_, err = container.GetProperties(d.Context, azblob.LeaseAccessConditions{})
 	if e, ok := err.(azblob.StorageError); ok {
 		if e.ServiceCode() == azblob.ServiceCodeContainerNotFound {
-			util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonContainerNotFound, "Container does not exist")
-			return false, nil
+			return nil
```
Comment (dmage) [inline pkg/storage/azure/azure.go]: The return value is ambiguous now. Does it mean "exists" or "doesn't exist"?

**Comment by coreydaley** [comment]
@dmage ptal, comments addressed
I decided to go the route of adding a boolean for whether it was found or not in addition to the error

**Comment by coreydaley** [comment]
/retest

[INFO] Acquiring a lease ...
[INFO] Lease acquired!
[INFO] Leased resource: {"type":"aws-quota-slice","name":"b99b9549-e9f7-4d4d-997f-adcd06c3a234","state":"leased","owner":"ci-op-i0fc0kqf-12abb","lastupdate":"2019-08-29T15:52:27.085648238Z","userdata":{}}
[INFO] Sending heartbeats to retain the lease...
failed to send heartbeat for resource "b99b9549-e9f7-4d4d-997f-adcd06c3a234": Post http://boskos.ci/update?name=b99b9549-e9f7-4d4d-997f-adcd06c3a234&owner=ci-op-i0fc0kqf-12abb&state=leased: dial tcp: lookup boskos.ci on 10.142.0.62:53: no such host
Another process exited
[INFO] Releasing the lease on resouce b99b9549-e9f7-4d4d-997f-adcd06c3a234...
failed to release resource "b99b9549-e9f7-4d4d-997f-adcd06c3a234": status 401 Unauthorized, statusCode 401 releasing b99b9549-e9f7-4d4d-997f-adcd06c3a234

**Comment by coreydaley** [comment]
@dmage @adambkaplan ptal
The e2e-azure-operator test failures look unrelated to this pull request.

File: pkg/storage/azure/azure.go [hunk 2]
```diff
@@ -358,19 +354,32 @@ func (d *driver) StorageExists(cr *imageregistryv1.Config) (bool, error) {
...
 	if err != nil {
-		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionUnknown, storageExistsReasonAzureError, fmt.Sprintf("Unable to get the storage container: %s", err))
-		return false, fmt.Errorf("unable to get the storage container %s: %s", d.Config.Container, err)
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonAzureError, fmt.Sprintf("%s", err))
+		return false, err
 	}
 
 	util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionTrue, storageExistsReasonContainerExists, "Storage container exists")
-	return true, nil
+	return exists, nil
```
Comment (adambkaplan) [inline pkg/storage/azure/azure.go]: Shouldn't the status condition reflect the `exists` return value?

File: pkg/storage/azure/azure.go [hunk 3]
```diff
@@ -310,44 +310,40 @@ func (d *driver) Secrets() (map[string]string, error) {
 	}, nil
 }
 
-// StorageExists checks if the storage container exists and accessible.
-func (d *driver) StorageExists(cr *imageregistryv1.Config) (bool, error) {
+// containerExists determines whether or not an azure container exists
+//
```
Comment (adambkaplan) [inline pkg/storage/azure/azure.go]: nit - extra comment line

File: pkg/storage/azure/azure.go [hunk 4]
```diff
@@ -495,6 +520,11 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 
 				break
 			}
+
+			if len(d.Config.Container) == 0 {
+				util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, "Unable to Generate Unique Bucket Name", "")
```
Comment (adambkaplan) [inline pkg/storage/azure/azure.go]:
```suggestion
				util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, "CreateStorageFailed", "Unable to generate a unique bucket name")
```

File: pkg/storage/azure/azure.go [hunk 5]
```diff
@@ -456,35 +465,51 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
...
+					generatedName = true
 				}
 
 				err = d.createStorageContainer(d.Config.AccountName, key, d.Config.Container)
 				if err != nil {
 					if e, ok := err.(azblob.StorageError); ok {
 						switch e.ServiceCode() {
 						case azblob.ServiceCodeContainerAlreadyExists:
+							if len(d.Config.Container) != 0 && !generatedName {
+								util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, "Unable to access container", "The container exists but we do not have permission to access it")
```
Comment (adambkaplan) [inline pkg/storage/azure/azure.go]: Convention for condition reason codes is `CamelCase` (and keep it short if we want these to bubble up to telemeter).

```suggestion
								util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, "StoragePermissionDenied", "The container exists but we do not have permission to access it")
```

File: pkg/storage/azure/azure.go (hunk 3 above)
Comment (coreydaley) [reply pkg/storage/azure/azure.go]: updated in next push

File: pkg/storage/azure/azure.go (hunk 1 above)
Comment (coreydaley) [reply pkg/storage/azure/azure.go]: updated in next push

**Comment by coreydaley** [comment]
@dmage @adambkaplan ptal, updates have been made per your suggestions

**Comment by coreydaley** [comment]
@dmage @adambkaplan bump

File: pkg/storage/azure/azure.go [hunk 6]
```diff
@@ -360,15 +355,33 @@ func (d *driver) StorageExists(cr *imageregistryv1.Config) (bool, error) {
...
+		return false, nil
+	}
+
+	exists, err := d.containerExists(d.Config.Container)
+	if err != nil {
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonAzureError, fmt.Sprintf("%s", err))
+		return false, err
+	}
+	if !exists {
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonContainerNotFound, "")
```
Comment (adambkaplan) [inline pkg/storage/azure/azure.go]: Provide a message if we couldn't find the storage container:
```suggestion
		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonContainerNotFound, fmt.Sprintf("Could not find storage container %s, d.Config.Container))
```

File: pkg/storage/azure/azure.go [hunk 7]
```diff
@@ -456,35 +469,50 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
...
+		var containerExists bool
+		if len(d.Config.Container) != 0 {
+			if containerExists, err = d.containerExists(d.Config.Container); err != nil {
+				util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonAzureError, fmt.Sprintf("%s", err))
+			}
+		}
+		if len(d.Config.Container) != 0 && containerExists {
+			cr.Status.Storage.Azure = d.Config.DeepCopy()
+			util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionTrue, storageExistsReasonContainerExists, "Azure container exists")
+		} else {
```
Comment (adambkaplan) [inline pkg/storage/azure/azure.go]: Can we simplify this a bit?

```
if containerExists {
   ...
   return nil
}

var generatedName bool
...
```

Comment (coreydaley) [reply pkg/storage/azure/azure.go]: updated in next push

File: pkg/storage/azure/azure.go (hunk 6 above)
Comment (coreydaley) [reply pkg/storage/azure/azure.go]: updated in next push

**Comment by coreydaley** [comment]
@adambkaplan ptal, comments have been addressed.

**Pull request merged**